sloth/toolBox.py

```diff
@@ -628,5 +632,119 @@ def mapDataRange_lin(X, y_min=0, y_max=1,
 
     return Y
 
+def intersection_calculations(df_data, corners, area_of_interest, nr_yr, nr_entries, save_dir):
+    """ Calculate spatial mean values for a shapefile in interest.
+
+    This function calculates spatial mean values (example for a specific region or province) taking into account how much (ratio) of the gridpoint is 
+    actually being intersected inside our region of interest.
+    
+    Parameters
+    -------------
+    df_data: dataframe
+         dataframe containing infromation about each gridpoint
+    
+    corners: dataframe (can also be a nectdf or any other data type)
+         containing infromation on the four longitudes and latitudes that surrounds each gridpoint in the source dataframe
+    
+    area_of_interest: a shapefile or geodataframe
+    	shapefile of the area of interest
+    
+    nr_yr: variable
+    	number of years of interest
+    
+    nr_entries:variable
+    	number of hours/days/or months etc..
+    
+    save_dir: path
+    	path for saving the output
+    
+    Returns
+    ---------
+    a shapefile of the area of interest including its corresponding spatial mean values
+    """
+
+    # create a geodataframe from the dataframe that we want to work with
+    df_data = pd.read_csv(df_data)
+    print("creating a geodataframe of the given dataframe")
+    gdf_data = gpd.GeoDataFrame(df_data, geometry=gpd.points_from_xy(df_data.lon,
+                                                                     df_data.lat))
+    ds_corners = xr.open_dataset(corners)
+
+    # convert the netcdf to dataframe in order to be able to convert it to geopandas and create polygons
+    df_corners = ds_corners.to_dataframe()
+    df_corners = df_corners.reset_index()
+    df_corners = df_corners.iloc[:, 4:]
+
+    # create a list of the four pairs (lon,lat) that creates the polygons
+    print("creating a list of four pairs (lon,lat) to create polygons")
+    list_poly = []
+    polygon_geom = list(zip(df_corners.grid_corner_lon, df_corners.grid_corner_lat))
+    for i in range(0, len(polygon_geom), 4):
+        list_poly.append(Polygon(polygon_geom[i:i + 4]))
+
+    # create a geodataframe from the polygon list that was created
+    print("creating a geodataframe with the polygons")
+    gdf_polygons = gpd.GeoDataFrame(geometry=list_poly)
+    gdf_polygons = gdf_polygons.drop_duplicates() #for some reason the polygons were doubled so a drop_duplicates() was used
+
+    # if the two goedataframes are do not have a coordinate systems, we have to set it first before going further setting the crs
+
+    gdf_data_wgs = gdf_data.set_crs('EPSG:4326') # setting the crs to WGS84
+    gdf_data_utm32N = gdf_data_wgs.to_crs('epsg:25832') # UTM N32 (This is in case of Germany, it will differ for other countries) 
```

Comment:
So this function is working for data over Germany only? If so, please do write a note in the doc-string of this function.